meta-mbl-distro/scripts/lib/mbl/update/appsimage.py

```diff
@@ -0,0 +1,152 @@
+# Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""PayloadImage subclass for apps images."""
+
+import arpy
+import pathlib
+import tarfile
+
+import mbl.update.payloadimage as upi
+import mbl.update.testinfo as testinfo
+import mbl.update.util as uutil
+
+MBL_APPS_ID = "APPS"
+
+
+class AppsImage(upi.PayloadImage):
+    """Class for creating image files containing apps (Opkg packages)."""
+
+    def __init__(self, app_paths):
+        """
+        Create an AppsImage object.
+
+        Args:
+        * app_paths list<str>: list of paths to the ipk files to add to the
+          payload.
+
+        """
+        self._apps = [uutil.ArchivedFileSpec(app) for app in app_paths]
+        _validate_app_paths(self._apps)
+
+    def stage(self, staging_dir):
+        """Implement method from PayloadImage ABC."""
+        upi.stage_multi_file_component(
+            staging_dir, self.archived_path, self._apps
+        )
+
+    def generate_testinfo(self):
+        """Implement method from PayloadImage ABC."""
+        return [
+            testinfo.app_bank_compare(_get_app_name(afs.path))
+            for afs in self._apps
+        ]
+
+    @property
+    def image_type(self):
+        """Implement method from PayloadImage ABC."""
+        return MBL_APPS_ID
+
+    @property
+    def image_format_version(self):
+        """Implement method from PayloadImage ABC."""
+        return 3
```

Comment:
Can't we use a global variable to store the version number?

Reply:
IIUC we can have a different version number for each image, so a global variable sounds like the wrong choice for this.

Reply:
We can have a global variable per component which at the moment have the same value. I don't like the idea of returning a fixed number in a function.

Reply:
Maybe we could have a class attribute in the base class and override it for each property in the concrete classes. However, I'm not sure it's worth the effort. These properties don't really behave like functions when called; they're intended to provide a layer of encapsulation for a class attribute you want as part of the public interface, so you can refactor or change the underlying class attribute without changing the public interface of the class. It's a little unusual to just return a fixed number, but python devs wouldn't find it too surprising as properties generally do just return "a value".
At the call site this would look something like `image.image_format_version == 3` so behaving more like an attribute than a function call.

Reply:
I agree with Rob. When we start having multiple versions of each `Image` class then it would make less sense to use globals.

Reply:
FWIW, when we bump versions, we'll be bumping the versions of each `Installer` class independently, and a single version that all the classes use won't work for that.

Before we do the next release, we'll need to support creating multiple versions of payloads within the script so that we can support updating from earlier versions of software over the air. At that point we'll have multiple versions of each image class so defining the version for a class outside the class itself will make less sense.

When we support multiple version of payload within the script then we'll need a richer mechanism for marking versions of classes and choosing the right one to use. I suggest we re-evaluate when implementing that mechanism.

Reply:
Rob and I decided that it would be better to embed the version numbers into the image types, so this code has changed: https://github.com/ARMmbed/meta-mbl/pull/763/commits/0953b4823cd59b89736faf359b51927654d3328f